apps/condo/domains/resident/tasks/helpers/sendResidentMessage.js

```diff
@@ -0,0 +1,181 @@
+const { isArray, isEmpty, compact, get, uniq } = require('lodash')
+
+const conf = require('@open-condo/config')
+const { getLogger } = require('@open-condo/keystone/logging')
+const { getSchemaCtx } = require('@open-condo/keystone/schema')
+const { i18n } = require('@open-condo/locales/loader')
+
+const { BillingProperty, BillingCategory } = require('@condo/domains/billing/utils/serverSchema')
+const { COUNTRIES } = require('@condo/domains/common/constants/countries')
+const { loadListByChunks } = require('@condo/domains/common/utils/serverSchema')
+const { MESSAGE_META } = require('@condo/domains/notification/constants/constants')
+const { sendMessage } = require('@condo/domains/notification/utils/serverSchema')
+const { Property } = require('@condo/domains/property/utils/serverSchema')
+const { Resident } = require('@condo/domains/resident/utils/serverSchema')
+
+const CHUNK_SIZE = 20
+const logger = getLogger('sendResidentMessage')
+
+/**
+ * Renders url from template by replacing {key} instances by data[key] values
+ * @param templateString
+ * @param data
+ * @returns {*}
+ */
+const renderTemplateString = (templateString, data) => {
+    const keys = Object.keys(data)
+    let result = templateString
+
+    for (const urlKey of keys) {
+        const value = get(data, urlKey)
+
+        if (value) result = result.replace(`{${urlKey}}`, value)
+    }
+
+    return result
+}
+
+/**
+ * Fills notification payload from request data according to message meta settings by selected notification type
+ * @param messageType
+ * @param notificationData
+ * @param data
+ */
+const fillExtraData = (messageType, notificationData) => {
+    const metaKeys = Object.keys(get(MESSAGE_META[messageType], 'data', {}))
+    const result = {}
+
+    for (const fieldKey of metaKeys) {
+        const value = get(notificationData, fieldKey)
+
+        if (value) result[fieldKey] = value
+    }
+
+    return result
+}
+
+/**
+ * Please have a look at SendResidentMessageService for json structure
+ * @param json
+ * @returns {Promise<{attemptsCount: (*), successCount: number}|{attemptsCount: number, successCount: number, json, error: Error}>}
+ */
+const sendResidentMessage = async (json) => {
+    let data
+
+    try {
+        data = JSON.parse(json)
+    } catch (error) {
+        logger.error({ msg: 'Error parsing JSON.', error: error.message })
+
+        return { attemptsCount: 0, successCount: 0, error, json }
+    }
+
+    const { organizationId, propertyDetails, uniqKeyTemplate, type: messageType, data: notificationData = {} } = data
+    const { title, message, urlTemplate, categoryId } = notificationData
+    const residentsWhere = { organization: { id: organizationId }, deletedAt: null }
+    const { keystone: context } = await getSchemaCtx('BillingProperty')
```

Comment:
you should not use any billing domain data for you task. Why do you really need it?

Reply:
EPS has only billingProperties info. thay're mapped to ordinary properties here, so it's needed actually.